Compiler/preproc/preprocessor.cpp

```diff
@@ -0,0 +1,429 @@
+
+    static const size_t NOT_FOUND = -1;
+
+#if AGS_PLATFORM_OS_WINDOWS
+    static const char * li_end = "\r\n";
+#else
+    static const char * li_end = "\n";
+#endif
+
+    class StringBuilder : TextWriter {
+    private:
+        String _str = nullptr;
+    public:
+        explicit StringBuilder(size_t size) {
+            _str.Reserve(size);
+        }
+
+        void WriteLine(const String& str) override { _str.Append(str); _str.Append(li_end); }
+        bool IsValid() const override { return true; }
+        void WriteChar(char c) override {  _str.AppendChar(c); }
+        void WriteString(const String &str) override { _str.Append(str);}
+        void WriteFormat(const char *fmt, ...) override { va_list argptr; _str.Append(String::FromFormatV(fmt, argptr)); }
+        void WriteLineBreak() override { _str.Append(li_end); }
+
+        String GetString () {return _str;}
+    };
+
+    class StringReader : TextReader{
+    private:
+        String _str = nullptr;
+        size_t _idx = 0;
+    public:
+        StringReader(const String& str) {
+            _str = str;
+            _idx = 0;
+        }
+
+        bool IsValid() const override { return _idx < _str.GetLength(); }
+        char ReadChar() override { return _str.GetAt(_idx++); }
+        String ReadString(size_t length) override { String res = _str.Mid(_idx, length); _idx+=res.GetLength(); return res;}
+        String ReadLine() override {
+            if(_idx >= _str.GetLength()) return nullptr;
+
+            size_t newPos = _str.FindChar('\n', _idx);
+            if(newPos != NOT_FOUND) {
+                size_t length = newPos - _idx;
+                size_t idx = _idx;
+                _idx = newPos+1;
+                if (length == 0) return String("");
+                return _str.Mid(idx,length);
+            } else {
+                return ReadAll();
+            }
+        }
+        String ReadAll() override { size_t idx = _idx; _idx = _str.GetLength(); return _str.Mid(idx, _str.GetLength() - idx); }
+    };
+
+    static bool Contains(std::vector<String> v, const String &s) {
+        for (std::vector<String>::reverse_iterator it = v.rbegin(); it != v.rend(); ++it ) {
+            if(s == *it) return true;
+        }
+        return false;
+    }
+
+
+    void Preprocessor::LogError(ErrorCode error, const String& message) {
+        if(message == nullptr) {
+            switch (error) {
+                case ErrorCode::MacroNameMissing:
+                    cc_error("Macro name expected");
+                    break;
+                case ErrorCode::LineTooLong:
+                    cc_error("Line too long (max line length = " STRINGIFY(MAX_LINE_LENGTH) ")");
+                    break;
+                case ErrorCode::EndIfWithoutIf:
+                    cc_error("#endif has no matching #if");
+                    break;
+                case ErrorCode::IfWithoutEndIf:
+                    cc_error("Missing #endif");
+                    break;
```

Comment:
I suggest adding a `default` case here, for asserting that a wrong error code is not passed in.

(Usually i just add "Unknown error" message to print something).

Reply:
Added the default here: https://github.com/adventuregamestudio/ags/pull/1370/commits/a922cee7f09b669c4b0073e75061cc14048ca579#diff-32411dcb2b7765703f7e4bc2ce6ca1f198d06ede7d9c7e2bcb02612b79c29615R93-R102

But notice not all messages go through this switch currently, I wanted to, but could not figure how to put all the messages in one place and deal with the arguments some error message take currently.